# Here we'll do a baseline test on 6 Neural Network model variations on the cleaned data

In [1]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
import psycopg2
import time

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import mean_squared_error
from pathlib import Path
import tensorflow as tf
import matplotlib.pyplot as plt

In [2]:
# Download raw data from postgres for stage 1 ETL

conn_string = 'postgres://whnpmxwsiccrtg:53c453893549d2b1e6a4ff92e626a2a08ebcaff66678e50d33e3742f66e3e4f4@ec2-52-4-171-132.compute-1.amazonaws.com/d2ajro4cjr10lb'

db = create_engine(conn_string)
conn = db.connect()

start_time = time.time()
df = pd.read_sql_query('select * from "merged_errors_corrected"',con=conn)
print("PostGres Download Duration: {} seconds".format(time.time() - start_time))
conn.close ()

PostGres Download Duration: 1.829629898071289 seconds


In [3]:
print(df.shape)
df.head()

(5553, 266)


host_listings_count  accommodates  bathrooms  bedrooms  price  \
0                    1             2        1.0         2   40.0   
1                    2             2        1.0         1   41.0   
2                    2             2        1.0         1   65.0   
3                    1             2        1.0         1   93.0   
4                    1             2        1.0         1  105.0   

   security_deposit  cleaning_fee  number_of_reviews  number_of_reviews_ltm  \
0               0.0           0.0                561                    131   
1               0.0          25.0                138                     22   
2               0.0          25.0                102                     20   
3               0.0           0.0                454                     53   
4             200.0           0.0                100                      1   

   review_scores_rating  ...  cancellation_policy_flexible  \
0                  99.0  ...                           0.0   
1                  94.0  ...                           0.0   
2                  96.0  ...                           0.0   
3                  97.0  ...                           0.0   
4                  99.0  ...                           0.0   

   cancellation_policy_moderate  cancellation_policy_strict  \
0                           0.0                         0.0   
1                           0.0                         0.0   
2                           0.0                         0.0   
3                           1.0                         0.0   
4                           0.0                         0.0   

   cancellation_policy_strict_14_with_grace_period  \
0                                              1.0   
1                                              1.0   
2                                              1.0   
3                                              0.0   
4                                              1.0   

   cancellation_policy_super_strict_30  cancellation_policy_super_strict_60  \
0                                  0.0                                  0.0   
1                                  0.0                                  0.0   
2                                  0.0                                  0.0   
3                                  0.0                                  0.0   
4                                  0.0                                  0.0   

   require_guest_profile_picture_t  require_guest_phone_verification_t  \
0                              0.0                                 0.0   
1                              1.0                                 1.0   
2                              1.0                                 1.0   
3                              1.0                                 1.0   
4                              1.0                                 1.0   

   has_availability_t  outlier  
0                 1.0        1  
1                 1.0        1  
2                 1.0        1  
3                 1.0        1  
4                 1.0        1  

[5 rows x 266 columns]

In [4]:
# Split our preprocessed data into our features and target arrays
y = df["price"].values
X = df.drop("price",1).values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [5]:
df.columns

Index(['host_listings_count', 'accommodates', 'bathrooms', 'bedrooms', 'price',
       'security_deposit', 'cleaning_fee', 'number_of_reviews',
       'number_of_reviews_ltm', 'review_scores_rating',
       ...
       'cancellation_policy_flexible', 'cancellation_policy_moderate',
       'cancellation_policy_strict',
       'cancellation_policy_strict_14_with_grace_period',
       'cancellation_policy_super_strict_30',
       'cancellation_policy_super_strict_60',
       'require_guest_profile_picture_t', 'require_guest_phone_verification_t',
       'has_availability_t', 'outlier'],
      dtype='object', length=266)

In [6]:
# scaler = RobustScaler()

from sklearn.preprocessing import RobustScaler
scaler = RobustScaler()

# Fit the scaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


# Test 1: Use one input layer and one hidden layer

In [7]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 80
hidden_nodes_layer2 = 30

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 80)                21280     
_________________________________________________________________
dense_1 (Dense)              (None, 30)                2430      
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 31        
Total params: 23,741
Trainable params: 23,741
Non-trainable params: 0
_________________________________________________________________


In [8]:
# Compile the model
nn.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

In [9]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
131/131 [==============================] - 1s 932us/step - loss: 199.0205 - mean_absolute_error: 199.0205
Epoch 2/50
131/131 [==============================] - 0s 895us/step - loss: 103.9814 - mean_absolute_error: 103.9814
Epoch 3/50
131/131 [==============================] - 0s 892us/step - loss: 90.9621 - mean_absolute_error: 90.9621
Epoch 4/50
131/131 [==============================] - 0s 905us/step - loss: 88.4032 - mean_absolute_error: 88.4032
Epoch 5/50
131/131 [==============================] - 0s 912us/step - loss: 84.7087 - mean_absolute_error: 84.7087
Epoch 6/50
131/131 [==============================] - 0s 899us/step - loss: 85.4314 - mean_absolute_error: 85.4314
Epoch 7/50
131/131 [==============================] - 0s 954us/step - loss: 84.3872 - mean_absolute_error: 84.3872
Epoch 8/50
131/131 [==============================] - 0s 1ms/step - loss: 81.1773 - mean_absolute_error: 81.1773
Epoch 9/50
131/131 [==============================] - 0s 944us/step - loss: 83

In [10]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}")
print(f"Mean Absolute Error: {model_accuracy}")

44/44 - 0s - loss: 59.8599 - mean_absolute_error: 59.8599
Loss: 59.85993194580078
Mean Absolute Error: 59.85993194580078


# Test 2.  Add more neurons to each layer

In [11]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 100
hidden_nodes_layer2 = 50

nn2 = tf.keras.models.Sequential()

# First hidden layer
nn2.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn2.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Output layer
nn2.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn2.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 100)               26600     
_________________________________________________________________
dense_4 (Dense)              (None, 50)                5050      
_________________________________________________________________
dense_5 (Dense)              (None, 1)                 51        
Total params: 31,701
Trainable params: 31,701
Non-trainable params: 0
_________________________________________________________________


In [12]:
# Compile the model
nn2.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model2 = nn2.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
131/131 [==============================] - 0s 887us/step - loss: 193.1178 - mean_absolute_error: 193.1178
Epoch 2/50
131/131 [==============================] - 0s 907us/step - loss: 97.9357 - mean_absolute_error: 97.9357
Epoch 3/50
131/131 [==============================] - 0s 855us/step - loss: 91.1235 - mean_absolute_error: 91.1235
Epoch 4/50
131/131 [==============================] - 0s 957us/step - loss: 87.6060 - mean_absolute_error: 87.6060
Epoch 5/50
131/131 [==============================] - 0s 851us/step - loss: 85.1315 - mean_absolute_error: 85.1315
Epoch 6/50
131/131 [==============================] - 0s 873us/step - loss: 87.9488 - mean_absolute_error: 87.9488
Epoch 7/50
131/131 [==============================] - 0s 844us/step - loss: 79.9202 - mean_absolute_error: 79.9202
Epoch 8/50
131/131 [==============================] - 0s 835us/step - loss: 81.1689 - mean_absolute_error: 81.1689
Epoch 9/50
131/131 [==============================] - 0s 832us/step - loss: 78

In [13]:
# Evaluate the model using the test data
model_loss2, model_accuracy2 = nn2.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss2}")
print(f"Mean Absolute Error: {model_accuracy2}")

44/44 - 0s - loss: 59.9744 - mean_absolute_error: 59.9744
Loss: 59.97441482543945
Mean Absolute Error: 59.97441482543945


# Test 3:  Add an additional layer with 20 neurons

In [14]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 100
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 20

nn3 = tf.keras.models.Sequential()

# First hidden layer
nn3.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn3.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Third hidden layer
nn3.add(tf.keras.layers.Dense(units=hidden_nodes_layer3,kernel_initializer='normal', activation="relu"))

# Output layer
nn3.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn3.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_6 (Dense)              (None, 100)               26600     
_________________________________________________________________
dense_7 (Dense)              (None, 50)                5050      
_________________________________________________________________
dense_8 (Dense)              (None, 20)                1020      
_________________________________________________________________
dense_9 (Dense)              (None, 1)                 21        
Total params: 32,691
Trainable params: 32,691
Non-trainable params: 0
_________________________________________________________________


In [15]:
# Compile the model
nn3.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model3 = nn3.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
131/131 [==============================] - 0s 1ms/step - loss: 185.8083 - mean_absolute_error: 185.8083
Epoch 2/50
131/131 [==============================] - 0s 924us/step - loss: 90.4279 - mean_absolute_error: 90.4279
Epoch 3/50
131/131 [==============================] - 0s 929us/step - loss: 87.3693 - mean_absolute_error: 87.3693
Epoch 4/50
131/131 [==============================] - 0s 899us/step - loss: 84.5110 - mean_absolute_error: 84.5110
Epoch 5/50
131/131 [==============================] - 0s 907us/step - loss: 79.9631 - mean_absolute_error: 79.9631
Epoch 6/50
131/131 [==============================] - 0s 967us/step - loss: 79.7438 - mean_absolute_error: 79.7438
Epoch 7/50
131/131 [==============================] - 0s 928us/step - loss: 78.5081 - mean_absolute_error: 78.5081
Epoch 8/50
131/131 [==============================] - 0s 898us/step - loss: 78.5220 - mean_absolute_error: 78.5220
Epoch 9/50
131/131 [==============================] - 0s 1ms/step - loss: 75.415

In [16]:
# Evaluate the model using the test data
model_loss3, model_accuracy3 = nn3.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss3}")
print(f"Mean Absolute Error: {model_accuracy3}")

44/44 - 0s - loss: 60.7458 - mean_absolute_error: 60.7458
Loss: 60.74576950073242
Mean Absolute Error: 60.74576950073242


# Test 4:  Add 50 neurons to the first layer and a 4th layer with 10 neurons

In [17]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 150
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 20
hidden_nodes_layer4 = 10

nn4 = tf.keras.models.Sequential()

# First hidden layer
nn4.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn4.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Third hidden layer
nn4.add(tf.keras.layers.Dense(units=hidden_nodes_layer3,kernel_initializer='normal', activation="relu"))

# Fourth hidden layer
nn4.add(tf.keras.layers.Dense(units=hidden_nodes_layer4,kernel_initializer='normal', activation="relu"))

# Output layer
nn4.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn4.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_10 (Dense)             (None, 150)               39900     
_________________________________________________________________
dense_11 (Dense)             (None, 50)                7550      
_________________________________________________________________
dense_12 (Dense)             (None, 20)                1020      
_________________________________________________________________
dense_13 (Dense)             (None, 10)                210       
_________________________________________________________________
dense_14 (Dense)             (None, 1)                 11        
Total params: 48,691
Trainable params: 48,691
Non-trainable params: 0
_________________________________________________________________


In [18]:
# Compile the model
nn4.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model4 = nn4.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
131/131 [==============================] - 0s 1ms/step - loss: 183.3469 - mean_absolute_error: 183.3469
Epoch 2/50
131/131 [==============================] - 0s 1ms/step - loss: 94.4479 - mean_absolute_error: 94.4479
Epoch 3/50
131/131 [==============================] - 0s 995us/step - loss: 85.9662 - mean_absolute_error: 85.9662
Epoch 4/50
131/131 [==============================] - 0s 1ms/step - loss: 83.1787 - mean_absolute_error: 83.1787
Epoch 5/50
131/131 [==============================] - 0s 1ms/step - loss: 79.2486 - mean_absolute_error: 79.2486
Epoch 6/50
131/131 [==============================] - 0s 995us/step - loss: 80.6275 - mean_absolute_error: 80.6275
Epoch 7/50
131/131 [==============================] - 0s 1ms/step - loss: 71.9787 - mean_absolute_error: 71.9787
Epoch 8/50
131/131 [==============================] - 0s 1ms/step - loss: 68.4172 - mean_absolute_error: 68.4172
Epoch 9/50
131/131 [==============================] - 0s 1ms/step - loss: 68.9822 - mean_a

In [19]:
# Evaluate the model using the test data
model_loss4, model_accuracy4 = nn4.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss4}")
print(f"Mean Absolute Error: {model_accuracy4}")

44/44 - 0s - loss: 59.3186 - mean_absolute_error: 59.3186
Loss: 59.31858444213867
Mean Absolute Error: 59.31858444213867


# Test 5:  Add 50 more neurons to the first layer and add a 5th layer with just 5 neurons

In [20]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 200
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 20
hidden_nodes_layer4 = 10
hidden_nodes_layer5 = 5

nn5 = tf.keras.models.Sequential()

# First hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="relu"))

# Third hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer3,kernel_initializer='normal', activation="relu"))

# Fourth hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer4,kernel_initializer='normal', activation="relu"))

# Fifth hidden layer
nn5.add(tf.keras.layers.Dense(units=hidden_nodes_layer5,kernel_initializer='normal', activation="relu"))


# Output layer
nn5.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn5.summary()

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_15 (Dense)             (None, 200)               53200     
_________________________________________________________________
dense_16 (Dense)             (None, 50)                10050     
_________________________________________________________________
dense_17 (Dense)             (None, 20)                1020      
_________________________________________________________________
dense_18 (Dense)             (None, 10)                210       
_________________________________________________________________
dense_19 (Dense)             (None, 5)                 55        
_________________________________________________________________
dense_20 (Dense)             (None, 1)                 6         
Total params: 64,541
Trainable params: 64,541
Non-trainable params: 0
__________________________________________________

In [21]:
# Compile the model
nn5.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model5 = nn5.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
131/131 [==============================] - 1s 1ms/step - loss: 194.5036 - mean_absolute_error: 194.5036
Epoch 2/50
131/131 [==============================] - 0s 1ms/step - loss: 99.8136 - mean_absolute_error: 99.8136
Epoch 3/50
131/131 [==============================] - 0s 1ms/step - loss: 84.2598 - mean_absolute_error: 84.2598
Epoch 4/50
131/131 [==============================] - 0s 1ms/step - loss: 83.1546 - mean_absolute_error: 83.1546
Epoch 5/50
131/131 [==============================] - 0s 1ms/step - loss: 88.0070 - mean_absolute_error: 88.0070
Epoch 6/50
131/131 [==============================] - 0s 1ms/step - loss: 77.3374 - mean_absolute_error: 77.3374
Epoch 7/50
131/131 [==============================] - 0s 1ms/step - loss: 80.1431 - mean_absolute_error: 80.1431
Epoch 8/50
131/131 [==============================] - 0s 1ms/step - loss: 79.1153 - mean_absolute_error: 79.1153
Epoch 9/50
131/131 [==============================] - 0s 1ms/step - loss: 78.1005 - mean_absol

In [22]:
# Evaluate the model using the test data
model_loss5, model_accuracy5 = nn5.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss5}")
print(f"Mean Absolute Error: {model_accuracy5}")

44/44 - 0s - loss: 59.7014 - mean_absolute_error: 59.7014
Loss: 59.70137023925781
Mean Absolute Error: 59.70137023925781


# Test 6:  Add 150 neurons to first layer and change second activation function to tanh

In [23]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_in_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 350
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 20
hidden_nodes_layer4 = 10
hidden_nodes_layer5 = 5

nn6 = tf.keras.models.Sequential()

# First hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer1,kernel_initializer='normal', input_dim=number_in_features, activation="relu"))

# Second hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer2,kernel_initializer='normal', activation="tanh"))

# Third hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer3,kernel_initializer='normal', activation="relu"))

# Fourth hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer4,kernel_initializer='normal', activation="relu"))

# Fifth hidden layer
nn6.add(tf.keras.layers.Dense(units=hidden_nodes_layer5,kernel_initializer='normal', activation="relu"))

# Output layer
nn6.add(tf.keras.layers.Dense(units=1, activation="linear"))

# Check the structure of the model
nn6.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_21 (Dense)             (None, 350)               93100     
_________________________________________________________________
dense_22 (Dense)             (None, 50)                17550     
_________________________________________________________________
dense_23 (Dense)             (None, 20)                1020      
_________________________________________________________________
dense_24 (Dense)             (None, 10)                210       
_________________________________________________________________
dense_25 (Dense)             (None, 5)                 55        
_________________________________________________________________
dense_26 (Dense)             (None, 1)                 6         
Total params: 111,941
Trainable params: 111,941
Non-trainable params: 0
________________________________________________

In [24]:
# Compile the model
nn6.compile(loss="mean_absolute_error", optimizer="adam", metrics=['mean_absolute_error'])

# Train the model
fit_model6 = nn6.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
131/131 [==============================] - 1s 1ms/step - loss: 223.9949 - mean_absolute_error: 223.9949
Epoch 2/50
131/131 [==============================] - 0s 1ms/step - loss: 184.5464 - mean_absolute_error: 184.5464
Epoch 3/50
131/131 [==============================] - 0s 1ms/step - loss: 113.1100 - mean_absolute_error: 113.1100
Epoch 4/50
131/131 [==============================] - 0s 1ms/step - loss: 83.5332 - mean_absolute_error: 83.5332
Epoch 5/50
131/131 [==============================] - 0s 1ms/step - loss: 83.2537 - mean_absolute_error: 83.2537
Epoch 6/50
131/131 [==============================] - 0s 1ms/step - loss: 75.6553 - mean_absolute_error: 75.6553
Epoch 7/50
131/131 [==============================] - 0s 1ms/step - loss: 74.4633 - mean_absolute_error: 74.4633
Epoch 8/50
131/131 [==============================] - 0s 1ms/step - loss: 69.3041 - mean_absolute_error: 69.3041
Epoch 9/50
131/131 [==============================] - 0s 2ms/step - loss: 63.4732 - mean_a

In [25]:
# Evaluate the model using the test data
model_loss6, model_accuracy6 = nn6.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss6}")
print(f"Mean Absolute Error: {model_accuracy6}")

44/44 - 0s - loss: 54.5875 - mean_absolute_error: 54.5875
Loss: 54.58751678466797
Mean Absolute Error: 54.58751678466797


# After lots of tests, our best Mean Squared Error Result is off, on average, by about $54.59 (according to the MAE)